packages/@glimmer/runtime/lib/vm/stack.ts

```diff
@@ -47,39 +42,41 @@ export class InnerStack {
     this.inner.copy(from, to);
   }
 
-  write(pos: number, value: unknown): void {
-    switch (typeof value) {
-      case 'boolean':
-      case 'undefined':
-        this.writeRaw(pos, encodeImmediate(value));
-        break;
-      case 'number':
-        if (isSmallInt(value)) {
-          this.writeRaw(pos, encodeImmediate(value));
-          break;
-        }
-      case 'object':
-        if (value === null) {
-          this.writeRaw(pos, encodeImmediate(value));
-          break;
-        }
-      default:
-        this.writeJs(pos, value);
-    }
-  }
-
   writeJs(pos: number, value: unknown): void {
     let idx = this.js.length;
     this.js.push(value);
     this.inner.writeRaw(pos, encodeHandle(idx));
   }
 
+  writeNumber(pos: number, value: number): void {
```

Comment:
this requires that things that call writeNumber guard isSmallInt not just typeof x === 'number' and there are a few places that are only doing the typeof check and calling pushNumber